### PR TITLE
Move database to `packages/graph`

### PR DESCRIPTION
**Pull request by TimDiekmann**
## 🌟 What is the purpose of this PR?

Setup basic infrastructure to run the Postgres database from the `packages/graph` folder.

## 🔗 Related links

- [Asana task](https://app.asana.com/0/0/1202532668544288/f) _(internal)_

## 🔍 What does this change?

- Create a docker setup for running the database in `packages/graph/deployment`: a647dcbbbd047d745474368710eed4005ffb70af
- Adjust Rust tests to pick up the new settings (small code quality improvements as drive-by as well): f968a8d9c4803737dc2ef091051a8794ed3a568c
- Rename the yarn commands from `yarn pg:...` to `yarn graph:...`

## 📜 Does this require a change to the docs?

A guide on how to start the database was added.

## ⚠️ Known issues

We currently don't have a final plan where everything will live, I put the readme at the top-level (`packages/graph`), so it's easy to pick up. This needs elaboration and a better structure later on

## 🐾 Next steps

Adjust the CI to ensure that the database is online when running the tests

## ❓ How to test this?

1.  Run the database
1.  Run `cargo test --workspace -- --ignored`
1.  Run `cargo run`

### CODE REVIEW

File: packages/graph/.gitignore
Comment (thehabbos007) [inline packages/graph/.gitignore:5]: I think we want to gitignore `.env.local`. Although it's already checked in/tracked, would be nice if it's still included but not tracked locally

Comment (TimDiekmann) [reply packages/graph/.gitignore:5]: Forgot to add this here as well, thank you!

Comment (thehabbos007) [reply packages/graph/.gitignore:5]: The `.env.local` needs to be there for the docker-compose to work (it references `.env.local`). We probably want to add to some readme to `touch .env.local` :)

File: packages/graph/hash_graph/bin/hash_graph/src/main.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/bin/hash_graph/src/main.rs:19]: I have looked into that and it wasn't as straightforward as I have hoped, so I'd leave that to a follow-up.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (benwerner01) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:131]: Just curious, was the removal of `.expect(...)` intentional here?

File: packages/graph/hash_graph/bin/hash_graph/src/main.rs
Comment (benwerner01) [reply packages/graph/hash_graph/bin/hash_graph/src/main.rs:19]: Cool, makes sense 👍

Comment (thehabbos007) [reply packages/graph/hash_graph/bin/hash_graph/src/main.rs:19]: @TimDiekmann I tried using [`dotenv`](https://crates.io/crates/dotenv) in the binary when I was testing out sqlx, does it not pick up the correct files, or are we not using that intentionally?

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:131]: Yes. I changed the function signature to return a `Result`, so it's possible to use the `?` operator instead of `.expect()`

File: packages/graph/hash_graph/bin/hash_graph/src/main.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/bin/hash_graph/src/main.rs:19]: The readme of `dotenv` states
> **Achtung!** This is a v0.* version! Expect bugs and issues all around. Submitting pull requests and issues is highly encouraged!

So I chose to not do that as a drive-by 🙃 
I guess this will not pick up `.env.local`.

File: packages/graph/.env
Comment (Alfred-Mountfield) [inline packages/graph/.env:5]: I don't mind this but why do we keep changing the name of the DB? From `postgres` to `postgres_graph` to `graph` now...

File: packages/graph/.gitignore
Comment (Alfred-Mountfield) [reply packages/graph/.gitignore:5]: > Although it's already checked in/tracked

I think if you add it to git then gitignoring it doesn't actually work how you want it to... At least it didn't last time I checked (even though I swear I used it like that before). If you add it to the gitignore, it'll still track changes if it's already been check-in.

File: packages/graph/README.md
Comment (Alfred-Mountfield) [inline packages/graph/README.md:11]:
```suggestion
1.  (Optional) If starting the database for the first time, the tables have to be generated. From the `migration` directory, run the following command:
```

File: packages/graph/README.md
Comment (Alfred-Mountfield) [inline packages/graph/README.md:17]:
```suggestion
1.  Then, apply the migrations from the `migration` directory:
```

Presumable this is going to apply all of them? (even if we have one at the moment) Or will it just apply the latest or something.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:13]:
```suggestion
    /// Creates a new `PostgresDatabase` connection.
```

Also I had object instead of connection because this isn't a connection... It's an object with a pool inside it and we'll likely have more things than a Pool.

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (Alfred-Mountfield) [inline packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:42]: I don't think this should block the PR but I think we should continue to think about how we use attachments, we didn't intend them to just be format strings as that's still embedding important information (like the URL) inside of plain-text, which means they're not programmatically accessible. We kind of wanted to be able to do `attach!(url=connect_url)` or something

**Review by Alfred-Mountfield — CHANGES_REQUESTED** [review]

File: packages/graph/.env
Comment (TimDiekmann) [reply packages/graph/.env:5]: We are already in a `postgres` database, I see no reason to use this in the database name, too.

File: packages/graph/README.md
Comment (TimDiekmann) [reply packages/graph/README.md:17]: I guess you are right. @benwerner01 can you confirm?

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs]: Changed this in 3faf90db86130a162128ce54c1ea63f65803dc92

File: packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs
Comment (TimDiekmann) [reply packages/graph/hash_graph/lib/graph/src/datastore/postgres/mod.rs:41]: I agree, that we probably shouldn't use `attach_printable` here. However, your proposed tracing-syntax doesn't work because we don't have a key (i.e. `url=`) here. I will make a few small adjustments and then we can decide if this is better.

**Review by Alfred-Mountfield — DISMISSED** [review]
LGTM 🚀

**Review by Alfred-Mountfield — APPROVED** [review]
Looks even more good to me 🚀

**Pull request merged**
